ENH: Add method to write GeoDataFrame to PostGIS

HTenkanen opened the pull request:
Closes #595, relates to #189, #440, #457, #546.

Adds a method `GeoDataFrame.to_postgis()` for writing data from GeoDataFrame to PostGIS.  

- Supports the same functionalities as Pandas, i.e. you can also use "replace" and "append" when pushing the GeoDataFrame to PostGIS. 
- When appending, the crs of the GeoDataFrame and the target table needs to match.

Example:
```
In [1]: data = gpd.read_file("https://gist.githubusercontent.com/HTenkanen/456ec4611a943955823a65729c9cf2aa/raw/be56f5e1e5c06c33cd51e89f823a7d770d8769b5/ykr_basegrid.geojson")
In [2]: engine = create_engine("postgresql+psycopg2://myuser:mypwd@localhost:5432/mydb")
In [3]: data.to_postgis(con=engine, name='test_table', if_exists='replace')
```

martinfleis commented on geopandas/geodataframe.py:
```diff
@@ -747,6 +747,40 @@ def astype(self, dtype, copy=True, errors="raise", **kwargs):
         return pd.DataFrame(df)
 
+    def to_postgis(
+        self, con, table, if_exists="fail", schema=None, dtype=None, index=False
```
Can we keep the same attribute names and order of attributes as pandas has in [to_sql](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.to_sql.html#pandas.DataFrame.to_sql)? For the sake of consistency. E.g. `table` should be `name` and order `self, name, con, schema=None, if_exists='fail', index=True, dtype=None`.

HTenkanen replied:
Thanks for your review @martinfleis! :) Sure, I'll check that this follows the same style as in Pandas.

HTenkanen replied:
Now we follow fully the pandas `to_sql` syntax including `chunksize` as well.

martinfleis commented on geopandas/geodataframe.py:
```diff
@@ -747,6 +747,40 @@ def astype(self, dtype, copy=True, errors="raise", **kwargs):
+        table : str
+            Target table to write the data.
+        if_exists : str
+            What to do if table exists already: 'replace' | 'append' | 'fail'.
```
You can copy the description from pandas as it is more explicit. May be the case for other as well.

martinfleis commented on geopandas/io/sql.py:
```diff
@@ -96,3 +96,222 @@ def load_geom_text(x):
+    geom_types = list(set(basic_types))
+
+    # Check for mixed geometry types
+    assert len(geom_types) < 2, "GeoDataFrame contains mixed geometry types."
```
Can you make a check and raise specific error instead of AssertionError to be more explicit?

```
if len(geom_types) < 2:
    raise...
```

HTenkanen replied:
Actually this "error" should not be an error at all but this should return GeometryCollection type. After diving more into the GeoAlchemy docs found that this indeed is possible..

HTenkanen replied:
If there is a mix of Geometry-types, they will be now inserted with PostGIS type `GEOMETRY`.

martinfleis commented on geopandas/io/sql.py:
```diff
@@ -96,3 +96,222 @@ def load_geom_text(x):
+    basic_types = []
+    for gt in geom_types:
+        if "Multi" in gt:
+            geom_collection = True
```
This does not check for GeometryCollection but Multi-types only. 

Btw, what would happen with GeometryCollections? Are they somehow supported?

HTenkanen replied:
Yep they indeed are supported, learned just now after your question (thanks!). I will add handling of those to here as well.

HTenkanen replied:
Added now support for GeometryCollections, as well as for LinearRings (which are converted to LineStrings).

martinfleis commented on geopandas/io/sql.py:
```diff
@@ -96,3 +96,222 @@ def load_geom_text(x):
+    """
+    Get EPSG code from CRS if available. If not, return -1.
+    """
+    from pyproj import CRS
```
Can you a TODO comment referring to #1101? This should be simplified once the other PR is merged.

HTenkanen replied:
Done.

martinfleis replied:
@HTenkanen Can be simplified now, pyproj.CRS is merged in master now, so gdf.crs will always be pyproj.CRS or None.

HTenkanen replied:
@martinfleis Excellent! I will update accordingly.

HTenkanen replied:
@martinfleis Okay, simplified now the codes accordingly and things are now read from the pyproj.CRS of the GeoDataFrame directly. Also added Geoalchemy as a dependency for the Python 3.8 conda-forge yaml.

martinfleis commented on geopandas/io/sql.py:
```diff
@@ -96,3 +96,222 @@ def load_geom_text(x):
+                srid = -1
+        except Exception:
+            srid = -1
+            print(
```
Can you use `warnings.warn` instead of pure print here?

martinfleis commented on geopandas/io/sql.py:
```diff
@@ -96,3 +96,222 @@ def load_geom_text(x):
     return GeoDataFrame(df, crs=crs, geometry=geom_col)
+
+
+def get_geometry_type(gdf):
```
```suggestion
def _get_geometry_type(gdf):
```
I would keep these helper functions private.

martinfleis commented on geopandas/io/sql.py:
```diff
@@ -96,3 +96,222 @@ def load_geom_text(x):
+    return (geom_type, geom_collection)
+
+
+def get_srid_from_crs(gdf):
```
```suggestion
def _get_srid_from_crs(gdf):
```

martinfleis commented on geopandas/io/sql.py:
```diff
@@ -96,3 +96,222 @@ def load_geom_text(x):
+    return srid
+
+
+def convert_to_wkb(gdf, geom_name):
```
```suggestion
def _convert_to_wkb(gdf, geom_name):
```

martinfleis commented on geopandas/io/sql.py:
```diff
@@ -96,3 +96,222 @@ def load_geom_text(x):
+    return gdf
+
+
+def write_to_db(gdf, engine, index, tbl, srid, geom_name, if_exists):
```
```suggestion
def _write_to_db(gdf, engine, index, tbl, srid, geom_name, if_exists):
```

martinfleis commented on geopandas/io/sql.py:
```diff
@@ -96,3 +96,222 @@ def load_geom_text(x):
+        cur.copy_expert(sql=sql, file=s_buf)
+
+        # SRID needs to be updated afterwards as the current approach does not support
+        # the use of EWKT geometries.
```
```suggestion
        # the use of WKT geometries.
```

HTenkanen replied:
Updated comment about EWKT/EWKB

martinfleis commented on geopandas/io/sql.py:
```diff
@@ -96,3 +96,222 @@ def load_geom_text(x):
+
+    except Exception as e:
+        conn.connection.rollback()
+        raise e
```
Do you know which Exception is expected? Can we provide some error message here to describe what happened?

HTenkanen replied:
There might be different errors happening when pushing data to the database. The error messages raised by psycopg2 are typically quite good, so I maybe would not deal them separately here.

martinfleis replied:
Alright, makes sense.

martinfleis commented on geopandas/io/sql.py:
```diff
@@ -96,3 +96,222 @@ def load_geom_text(x):
+
+
+def write_postgis(
+    gdf, con, table, if_exists="fail", schema=None, dtype=None, index=False
```
Names and order as above.

martinfleis commented on geopandas/io/sql.py:
```diff
@@ -96,3 +96,222 @@ def load_geom_text(x):
+
+    # Build target geometry type
+    if contains_multi_geoms:
+        target_geom_type = "Multi{geom_type}".format(geom_type=geometry_type)
```
Now I see that by _Geometry Collection_ above you meant Multi- geometry only. So GeometryCollections has to be somehow taken care of and I would prefer not using "collection" for Multi- as it is confusing :).

HTenkanen replied:
This is fixed in the latest version.

martinfleis commented on geopandas/tests/test_api.py:
```diff
@@ -18,6 +18,8 @@ def test_no_additional_imports():
         "psycopg2",
         "geopy",
+        "geoalchemy2",
+        "pygeos",
```
pygeos no longer used.
```suggestion

```

MulesNose commented:
Great to see these commits ! I have been trying to create my own workarounds to improve reading/writing to PostGIS from Geopandas, and have only made marginal gains. Looking forward to trying out these new improvements! Glad to see mixed geometry types (e.g. polygon and multi-) will be handled!

martinfleis commented on ci/travis/37-latest-conda-forge.yaml:
```diff
@@ -21,3 +21,6 @@ dependencies:
   - libspatialite
+  - geoalchemy2
+  # TEMP: workaround until conda-forge fixes the issues with gdal
+  - zstd
```
This was removed recently, can you delete it here as well?

HTenkanen replied:
@martinfleis Done!

martinfleis commented:
Thanks for all the updates! Looks good to me. AppVeyor fail is caused by Fiona toblerity/Fiona#851, so we should probably ignore it for now.

jorisvandenbossche commented on geopandas/io/sql.py:
```diff
@@ -96,3 +97,291 @@ def load_geom_text(x):
+    # Get geometry type and info whether data contains LinearRing
+    geometry_type, has_curve = _get_geometry_type(gdf)
+
+    # Build dtype with Geometry (srid is updated afterwards)
```
Why can't we specify the srid here when creating the type with geoalchemy?

HTenkanen replied:
Hmm, I'm not 100 % sure what you mean here? If you mean that why don't we use the `srid` parameter of [_GISType](https://geoalchemy-2.readthedocs.io/en/latest/types.html#geoalchemy2.types._GISType) object, it is because Shapely does not support EWKT or EWKB formats: https://community.gispython.narkive.com/qTVQCl3f/ewkt-ewkb-support-in-shapely

HTenkanen commented on geopandas/__init__.py:
```diff
@@ -4,6 +4,7 @@
 
 from geopandas.io.file import read_file  # noqa
 from geopandas.io.sql import read_postgis  # noqa
+from geopandas.io.sql import write_postgis  # noqa
```
Hi @jorisvandenbossche, and thanks for the review!

HTenkanen replied:
Agree, I'll remove this.

HTenkanen commented on geopandas/io/sql.py:
```diff
@@ -96,3 +97,291 @@ def load_geom_text(x):
+        if "Multi" in gt:
+            has_multi = True
+            # Keep track of the "root" geometry types
+            basic_types.append(gt.replace("Multi", ""))
```
@jorisvandenbossche Sorry for slow react on my side, have been quite busy. And thanks for pointing out this, indeed keeping track on root type is not necessary anymore. This was needed earlier before I realized we can use `GEOMETRY` geometry-type for storing mixtures of any geometries. I'll streamline this.

HTenkanen commented on geopandas/io/sql.py:
```diff
@@ -96,3 +97,291 @@ def load_geom_text(x):
+    from shapely.wkb import dumps
+
+    gdf[geom_name] = gdf[geom_name].apply(lambda geom: dumps(geom, hex=True))
+    return gdf
```
Cool thanks for pointing this out, I will add the hex to the `to_wkb`.

HTenkanen replied:
Done.

HTenkanen commented on geopandas/io/sql.py:
```diff
@@ -96,3 +97,291 @@ def load_geom_text(x):
+                has_curve = True
+        else:
+            has_single = True
+            basic_types.append(gt)
```
Yes, good point. 👍

HTenkanen commented on geopandas/geodataframe.py:
```diff
@@ -753,6 +753,54 @@ def astype(self, dtype, copy=True, errors="raise", **kwargs):
+            The keys should be the column names and the values
+            should be the SQLAlchemy types.
+        """
+        gdf = self.copy()
```
👍

HTenkanen commented:
@jorisvandenbossche Thanks for your review. I now updated most of the things according your comments.

Regarding the [Insertion method](https://pandas.pydata.org/pandas-docs/stable/user_guide/io.html#insertion-method), it took awhile to understand this. The docs in Pandas should perhaps be improved regarding this because it's not very clear how the `method` with callable should actually be used (a full example would be nice). 

Well, however, I was eventually able to grasp how it works, and now I implemented a version with that. You can check that from here: https://github.com/HTenkanen/geopandas/blob/insertion_method/geopandas/io/sql.py#L230-L236

However, I'm not sure if this is a bug in that insertion method, or maybe I still don't just understand how it should be used, but at the moment it seems that using the `pd_tbl.insert()` method with callable inserts the data always two times into the table? Any ideas @jorisvandenbossche or @martinfleis what could be going on in here?

martinfleis commented:
I have updated this PR to fit changes in master. However, note that all tests are currently skipped on CI and have been like that all the time (#1389).

HTenkanen commented:
@martinfleis Thanks for updating the PR! I will check locally if the tests pass.

martinfleis commented:
@HTenkanen we have now postgis running on CI in one of the envs. I have updated this PR to accommodate changes done in #1391, and two tests (using different schema) are failing. Can you check why is that?

HTenkanen commented:
@martinfleis Awesome! It's really good to have the db available in the environment :+1: I checked the error and the most likely error is that the db does not have the schema 'test' created which is the reason for fail. I'll fix the test so that it creates the schema beforehand. Probably might be good to catch this error and make a more user friendly error message as well. The one given by psycopg2 is not very intuitive.

HTenkanen commented:
@martinfleis Okay, fixed the test! Actually the error message from sqlalchemy was intuitive enough so I think it is not needed to create own error message regarding that for geopandas.

jorisvandenbossche commented on geopandas/io/tests/test_sql.py:
```diff
@@ -75,6 +76,95 @@ def connection_spatialite():
+    See create_postgis for more information.
+    """
+    sqlalchemy = pytest.importorskip("sqlalchemy")
+    from sqlalchemy.engine.url import URL
```
Can you add a `pytest.importorskip("psycopg2")` here?

HTenkanen replied:
Added this for engine creation fixture as well.

jorisvandenbossche commented on geopandas/io/tests/test_sql.py:
```diff
@@ -181,3 +271,333 @@ def test_read_postgis_binary(self, connection_spatialite, df_nybb):
+
+    def test_write_postgis_default(self, df_nybb):
+        """Tests that GeoDataFrame can be written to PostGIS with defaults."""
+        engine = connect_engine("test_geopandas")
+        if engine is None:
+            raise pytest.skip()
```
If we turn `connect_engine` into a fixture (like I did for `connection_postgis` now), then we can get rid of the skipping here, and also the `try/finally: engine.dispose()` below can be handled inside the fixture instead of needing to repeat here in every test.

HTenkanen replied:
Created now a fixture for engine as well and updated accordingly.

jorisvandenbossche commented on geopandas/io/tests/test_sql.py:
```diff
@@ -181,3 +271,333 @@ def test_read_postgis_binary(self, connection_spatialite, df_nybb):
+        table = "nybb"
+
+        try:
+            write_postgis(df_nybb, con=engine, name=table, if_exists="fail")
```
Can you first write in once with `if_exists="replace"` to ensure it certainly exists? (otherwise it might depend on the order in which tests are run whether this is actually testing what we want to test)

HTenkanen replied:
Done.

jorisvandenbossche commented on geopandas/io/tests/test_sql.py:
```diff
@@ -181,3 +271,333 @@ def test_read_postgis_binary(self, connection_spatialite, df_nybb):
+            # Validate
+            sql = "SELECT * FROM {table};".format(table=table)
+            df = read_postgis(sql, engine, geom_col="geometry")
+            validate_boro_df(df)
```
For some reason, this test is generating those warnings:

```
geopandas/io/tests/test_sql.py::TestIO::test_write_postgis_default
  /home/joris/miniconda3/envs/geo-dev/lib/python3.8/site-packages/sqlalchemy/dialects/postgresql/base.py:3027: SAWarning: Did not recognize type 'geometry' of column 'the_geom'
    util.warn(

geopandas/io/tests/test_sql.py::TestIO::test_write_postgis_default
  /home/joris/miniconda3/envs/geo-dev/lib/python3.8/site-packages/sqlalchemy/dialects/postgresql/base.py:3027: SAWarning: Did not recognize type 'geometry' of column 'geom'
    util.warn(

geopandas/io/tests/test_sql.py::TestIO::test_write_postgis_default
  /home/joris/miniconda3/envs/geo-dev/lib/python3.8/site-packages/sqlalchemy/dialects/postgresql/base.py:3027: SAWarning: Did not recognize type 'geometry' of column 'geometry'
    util.warn(
```

while the other tests are not .. But I don't really understand where this could be coming from

HTenkanen replied:
Yeah, I have been wondering about the same warnings as well, and don't know either what could be the reason for them.

jorisvandenbossche commented on geopandas/io/tests/test_sql.py:
```diff
@@ -75,6 +76,95 @@ def connection_spatialite():
+        metadata.reflect()
+        table = metadata.tables.get(table)
+        if table is not None:
+            base.metadata.drop_all(engine, [table], checkfirst=True)
```
```suggestion
            table.drop(checkfirst=True)
```

this seems to work for me as well (and then the `base = declarative_base()` is not needed)

HTenkanen replied:
Yes works for me as well. Updated accordingly.

jorisvandenbossche commented on geopandas/tests/util.py:
```diff
@@ -37,6 +37,34 @@ def validate_boro_df(df, case_sensitive=False):
 
 
+def connect_engine(dbname, user=None, password=None, host=None, port=None):
+    """
```
This is duplicating the version in test_sql.py?

HTenkanen replied:
Yeah indeed seemed to do so. I removed the one from `test_sql.py` and use the one from `util.py` now (also fixed it as it missed some imports).

HTenkanen replied:
Aaand let's take it back, just saw your comment about using the fixture. 👍

HTenkanen replied:
@jorisvandenbossche I guess we can now clean all the functions from `util.py` related to creating the connections?

jorisvandenbossche replied:
Yes, normally there shouldn't be any functions left in here that create a connection (the `create_..` functions below get a connection passed)

jorisvandenbossche commented on geopandas/io/sql.py:
```diff
@@ -98,3 +99,277 @@ def load_geom_text(x):
+
+    gdf = gdf.copy()
+    geom_name = gdf.geometry.name
+    if schema is not None:
+        schema_name = schema
+    else:
+        schema_name = "public"
```
Do you know whether this is needed? 
In the pandas sql code, no explicit schema name is used when not specified by the user (which I assume lets it to sqlalchemy to use the default schema)

HTenkanen replied:
Indeed this is only needed when parsing the SRID of the geometry column in existing table. I moved the codes under  `if_exists=="append"`. Why `schema_name` is needed relates to the use of this PostGIS function:
https://postgis.net/docs/Find_SRID.html

martinfleis commented:
> we have a problem with the recent CRS-related changes in `__setitem__` ... But it also makes me wonder a bit whether we should keep that warning in general

It is certainly not necessary there. It is there out of a convenience to help catch potential bug, but we may remove it. I am fine with that.

jorisvandenbossche commented:
@HTenkanen I gave the insertion method a try, and it *seems* to work (locally tests are passing). Please take a look at the last commit I pushed. If it works, it reduces quite a bit the custom code, as chunking, table creation/dropping if exists, etc is all handled by pandas now.

HTenkanen commented on geopandas/io/sql.py:
```diff
@@ -98,3 +99,277 @@ def load_geom_text(x):
+        try:
+            srid = gdf.crs.to_epsg(min_confidence=25)
+            if srid is None:
+                srid = -1
```
Good point! 👍 Added a warning.

HTenkanen commented on geopandas/io/sql.py:
```diff
@@ -98,3 +99,277 @@ def load_geom_text(x):
+
+    mask = gdf.geom_type == "LinearRing"
+    gdf.loc[mask, geom_name] = gdf.loc[mask, geom_name].apply(
+        lambda geom: LineString(geom)
```
Added a Todo comment about this.

HTenkanen commented on geopandas/io/sql.py:
```diff
@@ -98,3 +99,277 @@ def load_geom_text(x):
+    """Convert geometries to wkb. """
+    from geopandas.array import from_shapely, to_wkb
+
+    geom_array = from_shapely(gdf[geom_name])
+    gdf[geom_name] = to_wkb(geom_array, hex=True)
```
Updated accordingly.

HTenkanen replied:
Actually modified this further as I am now creating EWKB geometries.

HTenkanen commented on geopandas/io/sql.py:
```diff
@@ -98,3 +99,277 @@ def load_geom_text(x):
+        The keys should be the column names and the values
+        should be the SQLAlchemy types.
+    """
+    from geoalchemy2 import Geometry
```
Done 👍

HTenkanen replied:
Thanks @jorisvandenbossche for pointing out the possibility to create EWKB in shapely! :) Did some searches back in the days about this, but for some reason didn't ever find this issue about the topic. Hmm, I think it would definitely be a "cleaner" solution to create EWKB instead of doing these sql tricks with the geometry column. But let's check the influence on the performance and whether this also fixes the "append" method. I do some tests about these things.

HTenkanen replied:
@jorisvandenbossche Okay, did some tests and it seems that creating the EWKB takes twice as long compared to WKB without the srid information with shapely:

Creating 10 000 point wkbs without srid:
`116 ms ± 2.24 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)`

Creating 10 000 point ewkbs:
`226 ms ± 711 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)`

In case of updating the srid of pygeos array, I guess the cost is very slow as things are updated in a vectorized manner.  

Creating the EWKB geometries allows us to drop all of the SQL tricks related to updating the SRID of the geometry column. Hence, I would vote for accepting this cost increase in terms of shapely operations as the codes regarding the `to_postgis()` would become much leaner.

jorisvandenbossche commented:
> Okay, did some tests and it seems that creating the EWKB takes twice as long compared to WKB without the srid information with shapely:

And I did a quick timing with current pygeos, and here it seems the speed overhead is limited:

```python
# for 10000 points
In [1]: import pygeos 

In [2]: points = pygeos.points(np.random.randn(10000, 2)) 

In [3]: %timeit pygeos.to_wkb(points) 
4.8 ms ± 231 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [4]: %timeit pygeos.to_wkb(pygeos.set_srid(points, 4326), include_srid=True)   
5.83 ms ± 231 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

# for polygons/multipolygons
In [5]: df = geopandas.read_file("zip://./../Downloads/ne_10m_admin_1_states_provinces.zip") 

In [6]: geoms = df.geometry.values.data   

In [7]: %timeit pygeos.to_wkb(geoms)  
76.9 ms ± 966 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [8]: %timeit pygeos.to_wkb(pygeos.set_srid(geoms, 4326), include_srid=True)   
83.3 ms ± 1.53 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

(and things could even improve when implementing https://github.com/pygeos/pygeos/issues/137)


> Creating the EWKB geometries allows us to drop all of the SQL tricks related to updating the SRID of the geometry column. Hence, I would vote for accepting this cost increase in terms of shapely operations as the codes regarding the to_postgis() would become much leaner.

Yes, agreed.

jorisvandenbossche commented:
@HTenkanen while you are working on this, just want to make sure you saw that I pushed a commit yesterday (so that you pull that before pushing again)

HTenkanen commented:
@jorisvandenbossche Yes, I pulled the changes before starting to do these 👍 Thanks for updating the insert_copy thing, btw. It's now much better! 🙂

HTenkanen commented:
@jorisvandenbossche Okay, updated now the things according your comments. Now indeed the geometries are converted into EWKB before inserting to the database which allowed to streamline the codes quite a bit more now. I'm now using the pygeos directly under the `_to_ewkb()`. I think we might want to harmonize this and use the functions inside `array.py` and `_vectorized.py` but I now left things like this as I did not want to make too many changes at once before your review.

jorisvandenbossche commented on geopandas/io/tests/test_sql.py:
```diff
@@ -365,16 +364,12 @@ def test_write_postgis_append_when_table_exists(self, df_nybb):
 
         except AssertionError as e:
```
you can remove the try / except completely now, I think, since this except is just reraising the assertion error now

jorisvandenbossche replied:
(and similarly for some of the other tests)

HTenkanen replied:
Omg, unbelievable that I missed these 🙈 Well, good to have more eyes on the codes. Fixed these now for all tests.

jorisvandenbossche commented on geopandas/io/sql.py:
```diff
@@ -149,37 +151,53 @@ def _get_srid_from_crs(gdf):
+        from pygeos import from_shapely, set_srid, to_wkb
+
+        geoms = to_wkb(
+            set_srid(from_shapely(gdf[geom_name].values), srid=srid),
```
```suggestion
            set_srid(gdf[geom_name].values.data, srid=srid),
```

(when using pygeos, the values in the column are already pygeos geometries, so no `from_shapely` is needed)

HTenkanen replied:
Ah, still need some practice working with the pygeos version of gdp. Fixed this as well 👍

jorisvandenbossche commented:
@HTenkanen thanks a lot for the updates! Generally looks good, will do a final review later

jorisvandenbossche commented on geopandas/io/sql.py:
```diff
@@ -98,3 +101,227 @@ def load_geom_text(x):
+    else:
+        from shapely.wkb import dumps
+
+        geoms = [dumps(geom, srid=srid) for geom in gdf[geom_name]]
```
```suggestion
        geoms = [dumps(geom, srid=srid, hex=True) for geom in gdf[geom_name]]
```

this is causing the failures on travis

HTenkanen replied:
Oops, true. Just sent a fix.

HTenkanen commented:
> Now, Shapely actually can support creating EWKB, starting from shapely 1.7: Toblerity/Shapely#686 (comment) (and for older versions it could also be done by basically using the few lines of code how it is implemented in shapely). 

@jorisvandenbossche Regarding the above, what do you think would be the best way to proceed for older versions of Shapely? Should we implement and add our own `dumps` function somewhere and use that (not a fan of this option)? Or should we just raise an error and ask the user to update to a newer version of shapely?

jorisvandenbossche commented:
> Regarding the above, what do you think would be the best way to proceed for older versions of Shapely? Should we implement and add our own dumps function somewhere and use that (not a fan of this option)?

I think keeping a copy of `dumps` in our compat code for compatibility with older shapely versions is fine (it's certainly not the nicest option, but that's a quite typical way to deal with this, to include a "backport" of the feature in your code base) 
The shapely code: https://github.com/Toblerity/Shapely/blob/master/shapely/wkb.py#L24

(we should maybe discuss which shapely versions we want to support in the future)

jorisvandenbossche commented:
On second thought, maybe it's fine to simply require Shapely 1.7 for this. It's a new feature, people already need to have sqlalchemy and geoalchemy installed as well, and shapely 1.7 is already out now a few months.

HTenkanen commented:
@jorisvandenbossche Okay, either one works for me. :+1: I have now time to do the edits, so shall we go with raising the error then?

jorisvandenbossche commented:
> And another observation that I forgot to mention yesterday: It seems that using the new copy_insert method changes the order of how the rows are inserted into the database. 

That's strange .. I can't think of any reason why it would do that, we are still iterating over the rows and writing that into csv. 
Did you change something in one of the tests to get them passing for this? (in which case I can try to look at it if I can reproduce catching the failure)

jorisvandenbossche commented:
>  I have now time to do the edits, so shall we go with raising the error then?

Yes, that's good!

HTenkanen commented:
>Did you change something in one of the tests to get them passing for this? (in which case I can try to look at it if I can reproduce catching the failure)

Yes, I changed the  `test_write_postgis_mixed_geometry_types()` and `test_write_postgis_in_chunks()` to reflect the order that they were in the database. You can see the correct order how they should be in the db from the `df_mixed_single_and_multi()` decorator. Thanks for looking at this!

HTenkanen commented:
@jorisvandenbossche Okay, added now the import error if using `to_postgis()` with older shapely versions than '1.7.0'. In addition, realized that we hadn't considered 3D geometries, so added support for that as well.

HTenkanen commented:
@jorisvandenbossche By the way, do you know if pygeos/shapely 2.0 would start supporting 4D geometries (x, y, z, m)? Those are supported by PostGIS / GeoAlchemy2 so in principle it would be possible to write such geometries as well into the PostGIS table, but it seems that currently having 4 dimensions is not supported by neither shapely or PyGEOS.

jorisvandenbossche commented on geopandas/io/sql.py:
```diff
@@ -98,3 +101,257 @@ def load_geom_text(x):
+    geom_types = list(gdf.geometry.geom_type.unique())
+    has_curve = False
+
+    # Check for 3D-coordinates
+    dimension = 2
+    management = False
+    if any(gdf.geometry.has_z):
+        dimension = 3
+        management = True
```
I was looking at the geoalchemy2 docs (https://github.com/geoalchemy/geoalchemy2/blob/710719827e77a8c186ba863a0b200b869b4079df/geoalchemy2/types.py#L89-L91):

> With ``management`` set to ``False`` (the default) ``dimension`` is not
        taken into account, and the actual dimension is fully defined with the
        ``geometry_type``.

and it seems that they then add "Z" to the geometry type instead of using `dimension=3, management=True` (so using "GEOMETRYZ", "POINTZ", ..)

That might be the better option to avoid needing to do a "AddGeometryColumn" (which seems is not used anymore by default / recommended in postgis >= 2)

jorisvandenbossche replied:
So I think adding something like 

```
    if any(gdf.geometry.has_z):
        target_geom_type = target_geom_type + "Z"
```

should also do the trick

HTenkanen replied:
Ah, great! Let's do it that way :+1:

jorisvandenbossche commented:
> Yes, I changed the  `test_write_postgis_mixed_geometry_types()` and `test_write_postgis_in_chunks()` to reflect the order that they were in the database. You can see the correct order how they should be in the db from the `df_mixed_single_and_multi()` decorator. Thanks for looking at this!  

Hmm, this is still a bit strange. From looking into it, the change was not caused by the commit that changed the insertion method, but actually the one that changed to use EWKB (so where we no longer needed to `UpdateGeometrySRID` before and after copying to postgis, https://github.com/geopandas/geopandas/pull/1248/commits/b4cd568954a10edde33149d7b43dfe628f43952a). 
I checked out the commit adding the insertion method (https://github.com/geopandas/geopandas/pull/1248/commits/5eac3e0e7894338ad5cfb5461edf4638fb89b9c4), and then the tests are passing. When checking out the commit that changed the test, they are actually failing (as the order was still as original), when checking out the next commit which changed to use EWKB, then the tests are passing again.

Now, I am not sure this order is that important. It's only the order of `SELECT DISTINCT`, and I am not sure this is actually guaranteed to be following the order of the rows as inserted. 
If I do a `SELECT ST_AsText(geometry) FROM geomtype_tests;` in the middle of the test, it still returns the same order as the original GeoDataFrame:

```
test_geopandas=# SELECT ST_AsText(geometry) FROM geomtype_tests;
              st_astext               
--------------------------------------
 LINESTRING(0 0,1 1)
 MULTILINESTRING((0 0,1 1),(2 2,3 3))
 POINT(0 10)
(3 rows)

test_geopandas=# SELECT DISTINCT(GeometryType(geometry)) FROM geomtype_tests;
  geometrytype   
-----------------
 MULTILINESTRING
 POINT
 LINESTRING
(3 rows)
```

So I would maybe rather make the test more robust to order-changes, eg by doing `SELECT DISTINCT GeometryType(geometry) FROM geomtype_tests ORDER BY GeometryType(geometry);`

jorisvandenbossche commented:
@HTenkanen I pushed a commit for:

- the other option to support 2D geometries as mentioned above
- Ordering the select statement with multiple geometry types, so this is not dependent on the actual order returned by the database (why the order changed due to one of the commits, I have no idea, but I think it's not needed to figure this out)

HTenkanen commented:
@jorisvandenbossche Hah, was just about to push mine! 🙂 Sorry for a bit slow reaction.

jorisvandenbossche commented:
Whoops, sorry for the overlapping work. 

Do you have anything else remaining?

There is the warning when converting the geometries to WKB, but we can handle that separately in a follow-up.

> And by the way should we add test for checking that import error with older shapely versions? 

I think it is fine to leave it like this.

HTenkanen commented:
@jorisvandenbossche I added one more test to ensure that the row order indeed matches, but I'm not sure if that is actually needed as this thing should be catched on pandas side. Otherwise, I think this is it! 🎉 😎

jorisvandenbossche commented:
Feel free to push it!

HTenkanen commented:
@jorisvandenbossche Okay, added now the check for row order, and cleaned the test assertions a bit. I looked at the Pandas tests, and I'm not actually sure if there is a test that would specifically look at row order. 🤔 At least didn't find that right away. https://github.com/pandas-dev/pandas/blob/master/pandas/tests/io/test_sql.py

jorisvandenbossche commented:
There are roundtrip tests that use `assert_frame_equal`, so that also implies identical row order as the original dataframe.

jorisvandenbossche commented:
OK, let's get this in :tada: 

@HTenkanen thanks a lot!

HTenkanen commented:
@jorisvandenbossche Awesome!! 😄 🎉
